### ISSUE TITLE
Test command line tools with custom paths

### ISSUE THREAD

**Issue by clintongormley** [elastic/elasticsearch#12712]
Test commandline tools with custom paths in qa/
  - Especially stuff like custom plugin path

**Comment by colings86** [comment]
I ran the following as a manual test to test the beta and also to start recording tests that we might want to have for this when we write an automated test to cover it.

Installing a plugin to a custom path by running:
```
bin/plugin -Des.path.plugins=/path/to/custom/plugins install scirpt-python -u file:/path/to/elasticsearch-master/plugins/lang-python/target/releases/elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.zip
```

output on command line was:

```
-> Installing scirpt-python...
Trying file:/path/to/elasticsearch-master/plugins/lang-python/target/releases/elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.zip ...
Downloading ............................................................................................................................................................................................................................................................................................................................................DONE
PluginInfo{name='scirpt-python', description='The Python language plugin allows to have python as the language of scripts to execute.', site=false, jvm=true, classname=org.elasticsearch.plugin.python.PythonPlugin, isolated=true, version='2.0.0-beta1-SNAPSHOT'}
Installed scirpt-python into /path/to/custom/plugins/scirpt-python
```

`ls /path/to/custom/plugins/scirpt-python` gives: 
elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.jar jython-standalone-2.7.0.jar                        plugin-descriptor.properties

So it looks like installing a plugin to a custom path works

**Comment by clintongormley** [comment]
@colings86 probably need to test custom logs/config/scripts/repo paths as well

**Comment by colings86** [comment]
yup, getting to that now. Posting one test at a time ;)

**Comment by colings86** [comment]
### Listing plugins in a custom directory

Ran:
```
bin/plugin -Des.path.plugins=/path/to/custom/plugins list
```

output:
```
Installed plugins in /path/to/custom/plugins:
    - scirpt-python
```

SUCCESSFUL

### Removing a plugin using a custom plugins path

ran:
```
bin/plugin -Des.path.plugins=/path/to/custom/plugins remove scirpt-python
```

output:
```
-> Removing scirpt-python...
Removed scirpt-python
```

`/path/to/customplugins` directory no longer contained `scirpt-python` directory

SUCCESSFUL

**Comment by colings86** [comment]
## bin/plugin with custom config directory

For this test I created a conf directory at `/path/to/custom/config` and copied the default elasticsearch.yml file changing it only to add `path.plugins: /path/to/custom/plugins`

### Installing a plugin

ran:
```
bin/plugin -Des.path.conf=/path/to/custom/config install script-python -u file:/path/to/elasticsearch-master/plugins/lang-python/target/releases/elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.zip
```

output:
```
-> Installing script-python...
Trying file:/path/to/elasticsearch-master/plugins/lang-python/target/releases/elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.zip ...
Downloading ............................................................................................................................................................................................................................................................................................................................................DONE
PluginInfo{name='script-python', description='The Python language plugin allows to have python as the language of scripts to execute.', site=false, jvm=true, classname=org.elasticsearch.plugin.python.PythonPlugin, isolated=true, version='2.0.0-beta1-SNAPSHOT'}
log4j:WARN No appenders could be found for logger (org.elasticsearch.bootstrap).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Installed script-python into /path/to/custom/plugins/script-python
```

Confirmed `script-python` directory added to `/path/to/custom/plugins`

SUCCESSFUL

### Listing plugins

ran:
```
bin/plugin -Des.path.conf=/path/to/custom/config list
```

output:
```
Installed plugins in /path/to/custom/plugins:
    - script-python
```

SUCCESSFUL

### Remove plugins

ran:
```
bin/plugin -Des.path.conf=/path/to/custom/config remove script-python
```

output:
```
-> Removing script-python...
Removed script-python
```

Confirmed `script-python` directory removed from `/path/to/custom/plugins`

SUCCESSFUL

**Comment by colings86** [comment]
## bin/plugin with custom log directory

### Installing a plugin

ran:
```
bin/plugin -Des.path.logs=/path/to/custom/logs install script-python -u file:/path/to/elasticsearch-master/plugins/lang-python/target/releases/elasticsearch-lang-python-2.0.0-beta1-SNAPSHOT.zip
```

Confirmed `script-python` directory added to `/path/to/custom/plugins` and `/path/to/custom/logs` directory created with the following files:
```
elasticsearch.log                        elasticsearch_deprecation.log            elasticsearch_index_indexing_slowlog.log elasticsearch_index_search_slowlog.log
```

Also confirmed default logs directory was not created.

SUCCESSFUL

### Listing plugins

ran:
```
bin/plugin -Des.path.logs=/path/to/custom/logs list
```

output:
```
Installed plugins in /path/to/custom/plugins:
    - script-python
```

Confirmed `/path/to/custom/logs` directory created with the following files:
```
elasticsearch.log                        elasticsearch_deprecation.log            elasticsearch_index_indexing_slowlog.log elasticsearch_index_search_slowlog.log
```

Also confirmed default logs directory was not created.

SUCCESSFUL

### Remove plugins

ran:
```
bin/plugin -Des.path.logs=/path/to/custom/logs remove script-python
```

output:
```
-> Removing script-python...
Removed script-python
```

Confirmed `script-python` directory removed from `/path/to/custom/plugins` and `/path/to/custom/logs` directory created with the following files:
```
elasticsearch.log                        elasticsearch_deprecation.log            elasticsearch_index_indexing_slowlog.log elasticsearch_index_search_slowlog.log
```

Also confirmed default logs directory was not created.

SUCCESSFUL

**Comment by spinscale** [comment]
how exactly should this be implemented for automated tests? Would we have another project in `qa/` that installs/removes plugins with custom paths using custom `./integration-tests.xml` files?

Or do we have more `<ant antfile` calls in the `pom.xml` of the `smoke-test-plugins` that just installs plugins into different paths?

**Comment by brwe** [comment]
I made a pr with the first option (another project in qa/ that installs/removes plugins with custom paths) here: #12954.
I wonder if a bats test for this would be better though. Maybe @nik9000 has an opinion on this?

**Comment by nik9000** [comment]
> I wonder if a bats test for this would be better though. Maybe @nik9000 has an opinion on this?

Sorry I hadn't noticed this! Right now the bats tests are used to test `bin/plugin` and `bin/elasticsearch` with a few custom paths already so its not a huge stretch, especially when many path errors come from space escaping issues with bash. So its a reasonable choice.

**Comment by brwe** [comment]
This has been implemented as bats test now #12717, in https://github.com/elastic/elasticsearch/pull/13258 , https://github.com/elastic/elasticsearch/pull/13772 , https://github.com/elastic/elasticsearch/pull/13617 and https://github.com/elastic/elasticsearch/pull/13640
